src/foam/graphics/TreeNode.js

```diff
@@ -0,0 +1,297 @@
+        return neighborX + centerX;
+      },
+      postSet: function() {
+        this.neighborX = 0;
+        this.centerX = 0;
+        this.x = undefined;
+      }
+    },
+    {
+      class: 'Float',
+      name: 'neighborX'
+    },
+    {
+      class: 'Float',
+      name: 'centerX'
+    },
+    {
+      class: 'Boolean',
+      name: 'expanded',
+      expression: function(childNodes) {
+        return childNodes.length < 5;
+      },
+      postSet: function() {
+        this.graph.doLayout();
+      }
+    },
+    ['color', 'white']
+  ],
+
+  methods: [
+    function initCView() {
+      this.SUPER();
+
+      this.formatNode();
+
+      if (this.relationship) {
+        var data = this.data.clone(this.__subContext__);
+        try {
+          data[this.relationship.forwardName].select(data => {
+            this.addChildNode({ data: data });
+          });
+        } catch (x) { }
+        this.graph.doLayout();
+      }
+
+      this.canvas.on('click', (e) => {
+        var p = this.parent;
+
+        while ( this.cls_.isInstance(p) ) {
+          if ( ! p.expanded ) return;
+          p = p.parent;
+        }
+
+        var point = this.globalToLocalCoordinates({
+          w: 1,
+          x: e.offsetX,
+          y: e.offsetY,
+        });
+        point.x += this.width / 2;
+
+        var hit = this.hitTest(point)
+        if ( hit ) {
+          this.expanded = ! this.expanded;
+        }
+      });
+    },
+
+    function paint(x) {
+      if (!this.parentNode || this.parentNode.expanded) this.SUPER(x);
+    },
+
+    function paintSelf(x) {
+      x.save();
+
+      x.translate(-this.width / 2, 0);
+      this.SUPER(x);
+
+      // reset translate and shadow settings
+      x.restore();
+
+      this.paintConnectors(x);
+    },
+
+    function paintConnectors(x) {
+      function line(x1, y1, x2, y2) {
+        x.beginPath();
+        x.moveTo(x1, y1);
+        x.lineTo(x2, y2);
+        x.stroke();
+      }
+
+      x.lineWidth = 0.5; //this.borderWidth;
+      x.strokeStyle = this.border;
+
+      // Paint lines to childNodes
+      if (this.expanded && this.childNodes.length) {
+        var h = this.childNodes[0].y * 3 / 4;
+        var l = this.childNodes.length;
+
+        line(0, this.height, 0, h);
+        for (var i = 0; i < l; i++) {
+          var c = this.childNodes[i];
+          line(0, h, c.x, h);
+          line(c.x, h, c.x, c.y);
+        }
+      }
+
+      x.lineWidth = this.borderWidth;
+      // Paint expand/collapse arrow
+      if (this.childNodes.length) {
+        var d = this.expanded ? 5 : -5;
+        var y = this.height - 8;
+        line(-5, y, 0, y + d);
+        line(0, y + d, 5, y);
+      }
+    },
+
+    function addChildNode(args) {
+      var node = this.cls_.create(args, this);
+      node.y = this.height * 2;
+      this.add(node);
+      this.childNodes = this.childNodes.concat(node);
+      node.outline$.sub(() => {
+        this.outline = [];
```

Comment:
Explain why you need to do this. It's to force a recalculation of the outline because there's no way to subscribe to all children outlines via an expression.

Reply:
Added comment